src/main/java/picard/sam/util/SamComparison.java

```diff
@@ -1,67 +1,153 @@
+
+        if (!samComparisonArgumentCollection.LENIENT_DUP) {
+            throw new PicardException("Should only use countLenientDuplicateMarkingDifferences when in lenient duplicate marking mode.");
+        }
+
+        final DuplicateSetIterator duplicateSetLeftIterator = new DuplicateSetIterator(markDuplicatesCheckLeft.iterator(), leftReader.getFileHeader(), true);
+        final DuplicateSetIterator duplicateSetRightIterator = new DuplicateSetIterator(markDuplicatesCheckRight.iterator(), rightReader.getFileHeader(), true);
+        /* Iterate through duplicateSetIterators.  This will iterate through all reads with differing duplicate markings in left and right files.  Passing through the first
+         * iterator (i.e. left file), for each duplicate set we create a set of duplicate reads that the representative read could "swap" with.  On the pass through the second iterator, for each
+         * duplicate read, we look for the reads it is allowed to swap with in the duplicate set from the first iterator.  If the allowed swap can also be performed in this duplicate set,
+         * we do not count the read or the read it could be swapped with as having mismatched duplicates.  Note however, that once a swap is allowed, the reads swapped cannot be used for any
+         * further swaps.
+         */
+        final HashMap<String, Set<String>> allowedSwapsMap = new HashMap<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetLeftIterator)) {
+            // do not want to redo duplicate marking, want to keep marks assigned in input files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            if (records.size() > 1) {
+                final Set<String> duplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                        .collect(Collectors.toSet());
+                records.stream().filter(r -> !r.getDuplicateReadFlag())
+                        .forEach(r -> allowedSwapsMap.put(r.getReadName(), duplicateRecordNames));
+            }
+        }
+        // set of fragments which have been "saved", i.e. will not be counted as having mismatched duplicate marks.
+        final HashSet<String> savedNames = new HashSet<>();
+        for (final DuplicateSet duplicateSet : new IterableAdapter<>(duplicateSetRightIterator)) {
+            // do not want to redo duplicate marking, want to keep marking assigned in files
+            final List<SAMRecord> records = duplicateSet.getRecords(false);
+            if (records.size() > 1) {
+                final List<String> swappableDuplicateRecordNames = records.stream().filter(SAMRecord::getDuplicateReadFlag).map(SAMRecord::getReadName)
+                        .filter(allowedSwapsMap::containsKey).collect(Collectors.toList());
+                records.stream().filter(r -> !r.getDuplicateReadFlag()).map(SAMRecord::getReadName).filter(nRep -> !savedNames.contains(nRep))
```

Comment:
The parameter for this`filter` lambda:

`.filter(nRep -> !savedNames.contains(nRep)`

and for the `foreach` following it, are called `nRep` (which I read in this context to mean 'notRep' or 'nonRep'), but at this point its actually a stream of **representative** reads, isn't it (the previous filter only selects reads that don't have the duplicate flag set) ? Should these be called `repRead` or something ?